### PR TITLE
Support more OpenWRT TRX files by making fewer assumptions about the partitions

### PR DESCRIPTION
**Pull request by EdwardLarson**
**One sentence summary of this PR (This should go in the CHANGELOG!)**
Support more OpenWRT TRX files by making fewer assumptions about the partitions

**Link to Related Issue(s)**
#153
(#187, #161 previous PR)

**Please describe the changes in your request.**
The OpenWRT documentation [here](https://openwrt.org/docs/techref/headers) does not describe the TRX file format in the general case. The code [here](https://git.openwrt.org/?p=project/firmware-utils.git;a=blob_plain;f=src/otrx.c;hb=HEAD) shows how TRX files are actually created: up to three images (or partitions), possibly padded, are added to the file. These partitions do not necessarily correspond to loader, kernel, etc. described in the wiki or are not necessarily in that order. Therefore the TRX packer/unpacker needs to be much more general than it previously was.

**Anyone you think should look at this, specifically?**
@whyitfor @SamL98

### CODE REVIEW

File: ofrak_core/ofrak/core/openwrt.py [hunk 1]
```diff
@@ -412,65 +336,31 @@ class OpenWrtTrxPacker(Packer[None]):
     async def pack(self, resource: Resource, config=None):
         openwrt_v = await resource.view_as(OpenWrtTrx)
         header = await openwrt_v.get_header()
-        lzma_loader = await resource.get_only_child_as_view(
-            OpenWrtTrxLzmaLoader, ResourceFilter.with_tags(OpenWrtTrxLzmaLoader)
-        )
-        kernel = await resource.get_only_child_as_view(
-            OpenWrtTrxKernel, ResourceFilter.with_tags(OpenWrtTrxKernel)
+        children_by_offest = sorted(
```
Comment (EdwardLarson) [inline ofrak_core/ofrak/core/openwrt.py:339]:
```suggestion
        children_by_offset = sorted(
```

File: ofrak_core/ofrak/core/openwrt.py [hunk 2]
```diff
@@ -412,65 +336,31 @@ class OpenWrtTrxPacker(Packer[None]):
...
-        repacked_data_l.append(await kernel.resource.get_data())
-
-        trx_rootfs_offset = trx_kernel_offset + len(repacked_data_l[-1])
-        repacked_data_l.append(await rootfs.resource.get_data())
-
-        if header.get_version() == OpenWrtTrxVersion.VERSION2:
-            trx_binheader_offset = trx_rootfs_offset + len(repacked_data_l[-1])
-            repacked_data_l.append(await binheader.resource.get_data())
-
+        repacked_data_l = [await child.get_data() for _, child in children_by_offest]
```
Comment (EdwardLarson) [inline ofrak_core/ofrak/core/openwrt.py:347]:
```suggestion
        repacked_data_l = [await child.get_data() for _, child in children_by_offset]
```

File: ofrak_core/ofrak/core/openwrt.py [hunk 3]
```diff
@@ -412,65 +336,31 @@ class OpenWrtTrxPacker(Packer[None]):
...
-        else:
-            header_config = OpenWrtTrxHeaderModifierConfig(
-                trx_length=trx_length,
-                trx_crc=openwrt_crc32(data_to_crc),
-                trx_loader_offset=trx_loader_offset,
-                trx_kernel_offset=trx_kernel_offset,
-                trx_rootfs_offset=trx_rootfs_offset,
-                trx_binheader_offset=trx_binheader_offset,
-            )
+        offsets = [r.start for r, _ in children_by_offest]
```
Comment (EdwardLarson) [inline ofrak_core/ofrak/core/openwrt.py:351]:
```suggestion
        offsets = [r.start for r, _ in children_by_offset]
```

File: ofrak_core/test_ofrak/components/test_openwrt_component.py [hunk 4]
```diff
@@ -23,22 +23,30 @@ async def create_root_resource(self, ofrak_context: OFRAKContext) -> Resource:
...
 
     async def modify(self, resource: Resource) -> None:
         """
-        The modification for the OpenWrtTrx test is to replace the entirety of the kernel
+        The modification for the OpenWrtTrx test is to replace the entirety of the first partition
         with 4 null bytes
         """
-        kernel = await resource.get_only_child_as_view(
-            OpenWrtTrxKernel, ResourceFilter.with_tags(OpenWrtTrxKernel)
+        children_by_offest = sorted(
```
Comment (EdwardLarson) [inline ofrak_core/test_ofrak/components/test_openwrt_component.py:33]:
```suggestion
        children_by_offset = sorted(
```

File: ofrak_core/test_ofrak/components/test_openwrt_component.py [hunk 5]
```diff
@@ -23,22 +23,30 @@ async def create_root_resource(self, ofrak_context: OFRAKContext) -> Resource:
...
+                (await child.get_data_range_within_root(), child)
+                for child in await resource.get_children()
+                if not child.has_tag(OpenWrtTrxHeader)
+            ],
+            key=lambda x: x[0].start,
         )
-        original_size = await kernel.resource.get_data_length()
-        kernel.resource.queue_patch(Range(0, original_size), b"\x00\x00\x00\x00")
-        await kernel.resource.save()
+        partition = children_by_offest[0][1]
```
Comment (EdwardLarson) [inline ofrak_core/test_ofrak/components/test_openwrt_component.py:41]:
```suggestion
        partition = children_by_offset[0][1]
```

**Comment by EdwardLarson** [comment]
LGTM! (though this is technically my PR, none of the code is mine, so I count my review as a proper review)

**Review by whyitfor — APPROVED** [review]

**Pull request merged**